sdk/core/azure-core/src/http/url.cpp

```diff
@@ -1,69 +1,271 @@
+    char c = value[i];
+    if (c == '+')
+    {
+      decodedValue += ' ';
+      ++i;
+    }
+    else if (c == '%')
     {
-      if (this->m_path[pathLast - index] != '/')
+      if (i + 2 >= value.size() || hexTable[value[i + 1]] < 0 || hexTable[value[i + 2]] < 0)
       {
-        this->m_path = this->m_path.substr(0, pathSize - index);
-        break;
+        throw std::runtime_error("failed when decoding url component");
       }
+      int v = (hexTable[value[i + 1]] << 4) + hexTable[value[i + 2]];
+      decodedValue += static_cast<std::string::value_type>(v);
+      i += 3;
+    }
+    else
+    {
+      decodedValue += value[i];
+      ++i;
+    }
+  }
+  return decodedValue;
+}
+
+static const char* unreserved
+    = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~";
+static const char* subdelimiters = "!$&'()*+,;=";
+
+std::string Url::EncodeHost(const std::string& host)
+{
+  return EncodeImpl(host, [](int c) { return c > 127; });
+}
+
+std::string Url::EncodePath(const std::string& path)
+{
+  const static std::vector<bool> shouldEncodeTable = []() {
+    const std::string pathCharacters
+        = std::string(unreserved) + std::string(subdelimiters) + "%/:@";
+
+    std::vector<bool> ret(256, true);
+    for (char c : pathCharacters)
+    {
+      ret[c] = false;
+    }
+    // we also encode % and +
+    ret['%'] = true;
+    ret['+'] = true;
+    return ret;
+  }();
+
+  return EncodeImpl(path, [](int c) { return shouldEncodeTable[c]; });
+}
+
+std::string Url::EncodeQuery(const std::string& query)
+{
+  const static std::vector<bool> shouldEncodeTable = []() {
+    std::string queryCharacters = std::string(unreserved) + std::string(subdelimiters) + "%/:@?";
+
+    std::vector<bool> ret(256, true);
+    for (char c : queryCharacters)
+    {
+      ret[c] = false;
+    }
+    // we also encode % and +
+    ret['%'] = true;
+    ret['+'] = true;
+    // Surprisingly, '=' also needs to be encoded because Azure Storage server side is so strict.
+    // We are applying this function to query key and value respectively, so this won't affect
+    // that = used to separate key and query.
+    ret['='] = true;
+    return ret;
+  }();
+
+  return EncodeImpl(query, [](int c) { return shouldEncodeTable[c]; });
+}
+
+std::string Url::EncodeFragment(const std::string& fragment)
+{
+  const static std::vector<bool> shouldEncodeTable = []() {
+    std::string queryCharacters = std::string(unreserved) + std::string(subdelimiters) + "%/:@?";
+
+    std::vector<bool> ret(256, true);
+    for (char c : queryCharacters)
+    {
+      ret[c] = false;
+    }
+    // we also encode % and +
+    ret['%'] = true;
+    ret['+'] = true;
+    return ret;
+  }();
+
+  return EncodeImpl(fragment, [](int c) { return shouldEncodeTable[c]; });
+}
+
+std::string Url::EncodeImpl(const std::string& source, const std::function<bool(int)>& shouldEncode)
+{
+  const char* hex = "0123456789ABCDEF";
```

Comment:
Just a question, why did we pick upper case instead of lower case?

Reply:
as suggested in RFC


```
For consistency, URI producers and
   normalizers should use uppercase hexadecimal digits for all percent-
   encodings.
```

Reply:
All this is valuable information/explanation that we should put as code comments, with links to the relevant RFC section.